runtime/src/iree/hal/utils/libmpi.h

```diff
@@ -0,0 +1,77 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef IREE_HAL_UTILS_LIBMPI_H_
+#define IREE_HAL_UTILS_LIBMPI_H_
+
+#include "iree/base/internal/dynamic_library.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif  // __cplusplus
+
+// A collection of known MPI symbols (functions and handles)
+//
+// To support additional functions, define the function or handles
+// in libmpi_dynamic_symbols.h
+typedef struct iree_hal_mpi_dynamic_symbols_t {
+#define MPI_PFN_DECL(mpiSymbolName, ...) int (*mpiSymbolName)(__VA_ARGS__);
+#include "iree/hal/utils/libmpi_dynamic_symbols.h"
+#undef MPI_PFN_DECL
+} iree_hal_mpi_dynamic_symbols_t;
+
+// Converts a mpi result to an iree_status_t.
+//
+// Usage:
+//   iree_status_t status = MPI_RESULT_TO_STATUS(mpiDoThing(...));
+#define MPI_RESULT_TO_STATUS(syms, expr, ...) \
+  iree_hal_mpi_result_to_status((syms), ((syms)->expr), __FILE__, __LINE__)
+
+// Converts a mpi result to a Status object.
+iree_status_t iree_hal_mpi_result_to_status(
+    iree_hal_mpi_dynamic_symbols_t* syms, int result, const char* file,
+    uint32_t line);
+
+// IREE_RETURN_IF_ERROR but implicitly converts the mpi return value to
+// a Status.
+//
+// Usage:
+//   MPI_RETURN_IF_ERROR(mpiDoThing(...), "message");
+#define MPI_RETURN_IF_ERROR(syms, expr, ...)                                 \
+  IREE_RETURN_IF_ERROR(iree_hal_mpi_result_to_status((syms), ((syms)->expr), \
+                                                     __FILE__, __LINE__),    \
+                       __VA_ARGS__)
+
+// IREE_IGNORE_ERROR but implicitly converts the mpi return value to a
+// Status.
+//
+// Usage:
+//   MPI_IGNORE_ERROR(mpiDoThing(...));
+#define MPI_IGNORE_ERROR(syms, expr)                                      \
+  IREE_IGNORE_ERROR(iree_hal_mpi_result_to_status((syms), ((syms)->expr), \
+                                                  __FILE__, __LINE__))
+
+// Dynamically loads libmpi and sets up the symbol table.
+//
+// Args:
+//  - host_allocator: the allocator to use to create the library and symbol
+//    objects.
+//  - out_library: out parameter -- on success, it holds a pointer to the
+//    dynamic library object.
+//  - out_syms: out parameter -- on success, it populates the list of symbols
+//    available from the MPI library
```

Comment:
Style: https://google.github.io/styleguide/cppguide.html#Function_Comments

We generally only include relevant information about arguments (nearly all functions take an iree_allocator_t so it's elided, etc) - the `out_` naming of args also covers the behavior such that it can be elided. The args themselves are usually referenced by `|arg|` (at least back in the day that was google style, backticks may be sufficient now but we'd want to do a big find/replace to make them consistent). Keeping the description short is useful for quick referencing in tools that show the first few lines of documentation:
![image](https://user-images.githubusercontent.com/75337/234464006-421c1c52-1374-477d-ba27-3a0c2f71d46d.png)

So this could read:
```suggestion
// Dynamically loads MPI and initializes the symbol table.
// |out_library| must be released by the caller with iree_dynamic_library_release.
// |out_syms| is only valid for as long as the dynamic library is live.
```
and convey the same information plus the useful bits about lifetime/ownership.

Reply:
Thanks! done.